contracts/src/libraries/IGasRefunder.sol

```diff
@@ -21,14 +21,20 @@ abstract contract GasRefundEnabled {
         uint256 startGasLeft = gasleft();
         _;
         if (address(gasRefunder) != address(0)) {
-            uint256 calldataSize = 0;
+            uint256 calldataSize;
+            assembly {
+                calldataSize := calldatasize()
+            }
+            uint256 calldataWords = (calldataSize + 31) / 32;
+            // account for the CALLDATACOPY cost of the proxy contract, including the memory expansion cost
+            startGasLeft += calldataWords * 6 + (calldataWords**2) / 512;
             // if triggered in a contract call, the spender may be overrefunded by appending dummy data to the call
             // so we check if it is a top level call, which would mean the sender paid calldata as part of tx.input
             // solhint-disable-next-line avoid-tx-origin
-            if (msg.sender == tx.origin) {
-                assembly {
-                    calldataSize := calldatasize()
-                }
+            if (msg.sender != tx.origin) {
+                // We can't be sure if this calldata came from the top level tx,
+                // so to be safe we tell the gas refunder there was no calldata.
+                calldataSize = 0;
```

Comment:
It'd be really nice to have a test for this

Reply:
I'm not sure exactly how we'd test this -- we could create a mock gas refunder but that seems like a lot of work for a test of this condition

Reply:
Was just looking back at this PR, I had really meant a test making sure the non-constant costs were covered rather than a test of this condition